src/Pate/Verification.hs

```diff
@@ -929,110 +533,546 @@ regStateToAsn regs = do
+    univDom <- universalDomainSpec
+    proveLocalPostcondition bundle isUnknown univDom
+
+  let allPreconds = precondReturn:precondUnknown:preconds
+
+  checkCasesTotal bundle allPreconds
+  foldM (\stPred (p, (_, stPred')) -> liftIO $ muxStatePred sym p stPred' stPred) (statePredFalse sym) allPreconds
+
+matchingExits ::
+  forall sym arch.
+  SimBundle sym arch ->
+  MS.MacawBlockEndCase ->
+  EquivM sym arch (W4.Pred sym)
+matchingExits bundle ecase = withSym $ \sym -> do
+  case1 <- liftIO $ MS.isBlockEndCase (Proxy @arch) sym (simOutBlockEnd $ simOutO bundle) ecase
+  case2 <- liftIO $ MS.isBlockEndCase (Proxy @arch) sym (simOutBlockEnd $ simOutP bundle) ecase
+  liftIO $ W4.andPred sym case1 case2  
+
+-- | Prove a local postcondition (i.e. it must hold when the slice exits) for a pair of slices
+proveLocalPostcondition ::
+  HasCallStack =>
+  SimBundle sym arch ->
+  W4.Pred sym ->
+  StatePredSpec sym arch ->
+  EquivM sym arch (W4.Pred sym, StatePred sym arch)
+proveLocalPostcondition bundle branchCase postcondSpec = withSym $ \sym -> do
+  eqRel <- asks envBaseEquiv
+  (asm, postcond) <- liftIO $ bindSpec sym (simOutState $ simOutO bundle) (simOutState $ simOutP bundle) postcondSpec
+  -- this weakened equivalence relation is only used for error reporting
+  (postEqRel, postcondPred) <- liftIO $ getPostcondition sym bundle eqRel postcond
+  
+  fullPostCond <- liftIO $ allPreds sym [postcondPred, branchCase]
+  eqInputs <- withAssumption_ (return asm) $
+    guessEquivalenceDomain bundle fullPostCond postcond
+
+  stackRegion <- asks envStackRegion
+  eqInputsPred <- liftIO $ getPrecondition sym stackRegion bundle eqRel eqInputs
+
+  notChecks <- liftIO $ W4.notPred sym fullPostCond
+  (oBlocks, pBlocks) <- getBlocks bundle
+
+  startedAt <- liftIO TM.getCurrentTime
+  result <- withAssumption_ (liftIO $ allPreds sym [eqInputsPred, asm]) $ do
+    checkSatisfiableWithModel "check" notChecks $ \satRes -> do        
+      finishedBy <- liftIO TM.getCurrentTime
+      let duration = TM.diffUTCTime finishedBy startedAt
+
+      case satRes of
+        W4R.Unsat _ -> do
+          emitEvent (PE.CheckedEquivalence oBlocks pBlocks PE.Equivalent duration)
+          return Nothing
+        W4R.Unknown -> do
+          emitEvent (PE.CheckedEquivalence oBlocks pBlocks PE.Inconclusive duration)
+          throwHere InconclusiveSAT
+        W4R.Sat fn -> do
+          ir <- getInequivalenceResult InvalidPostState postEqRel bundle fn
+          emitEvent (PE.CheckedEquivalence oBlocks pBlocks (PE.Inequivalent ir) duration)
+          return $ Just ir
+        
+  case result of
+    Just result' -> throwHere $ InequivalentError result'
+    _ -> return (eqInputsPred, eqInputs)
+
+--------------------------------------------------------
+-- Propagating preconditions
+
+-- | Guess a sufficient memory domain that will cause the
+-- given postcondition to be satisfied on the given equivalence relations
```

Comment:
What sorts of things go into a guess? Also, the guess is just used as an optimization and is not required to be correct?